modules/500-upmeter/hooks/generate_password.go

```diff
@@ -17,127 +17,127 @@ limitations under the License.
 package hooks
 
 import (
-	"encoding/json"
 	"fmt"
+	"strings"
 
 	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
 	"github.com/flant/addon-operator/sdk"
+	"github.com/flant/shell-operator/pkg/kube_events_manager/types"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 
 	"github.com/deckhouse/deckhouse/go_lib/pwgen"
 )
 
 var _ = sdk.RegisterFunc(&go_hook.HookConfig{
-	Queue:        "/modules/upmeter/generate_password",
+	Queue: "/modules/upmeter/generate_password",
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:       authSecretBinding,
+			ApiVersion: "v1",
+			Kind:       "Secret",
+			NameSelector: &types.NameSelector{
+				MatchNames: authSecretNames,
+			},
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{upmeterNS},
+				},
+			},
+			// Synchronization is redundant because of OnBeforeHelm.
+			ExecuteHookOnSynchronization: go_hook.Bool(false),
+			ExecuteHookOnEvents:          go_hook.Bool(false),
+			FilterFunc:                   filterAuthSecret,
+		},
+	},
+
 	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10},
-}, generatePassword)
+}, restoreOrGeneratePassword)
+
+const (
+	upmeterNS         = "d8-upmeter"
+	authSecretField   = "auth"
+	authSecretBinding = "auth-secrets"
+	statusSecretName  = "basic-auth-status"
+	webuiSecretName   = "basic-auth-webui"
+
+	externalAuthValuesTmpl     = "upmeter.auth.%s.externalAuthentication"
+	passwordValuesTmpl         = "upmeter.auth.%s.password"
+	passwordInternalValuesTmpl = "upmeter.internal.auth.%s.password"
+)
 
-func generatePassword(input *go_hook.HookInput) error {
-	const rootAuthKey = "upmeter.auth"
+var authSecretNames = []string{statusSecretName, webuiSecretName}
+var upmeterApps = map[string]string{
+	statusSecretName: "status",
+	webuiSecretName:  "webui",
+}
+
+type storedPassword struct {
+	SecretName string `json:"name"`
+	Password   string `json:"password,omitempty"`
+}
 
-	values, config, err := parseValuesAndConfig(rootAuthKey, input)
+func filterAuthSecret(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	secret := &v1.Secret{}
+	err := sdk.FromUnstructured(obj, secret)
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("cannot convert secret to struct: %v", err)
 	}
 
-	config.Status = setPassword(config.Status, values.Status)
-	config.Webui = setPassword(config.Webui, values.Webui)
-
-	if config.IsEmpty() {
-		input.ConfigValues.Remove(rootAuthKey)
-		return nil
+	password := string(secret.Data[authSecretField])
```

Comment:
I'd say it is credentials, not just password

Reply:
It become a password later after trimming ;)